### PR TITLE
Microsoft Bot Framework Publish Validation for .NET

### PR DESCRIPTION
**Pull request by enzocanoo**
This tool is used for validating a Bot's Configuration prior to publishing it, in order to prevent common issues like missing endpoints.

There are several validation checks available at the moment. These are the following:
- **ForbidSpacesInProjectName**: No spaces in the Project's name (this is due to a [bug in .NET/Azure](https://github.com/aspnet/websdk/issues/237))
- **RequireBotFile**: Existence of a '.bot' file
- Internal structure of the configuration file:
    - **RequireEndpoints**: Required endpoints  
    - **ForbidEndpoints**: Forbidden endpoints
    - **RequireLuisKey**: Existence of LUIS key
    - **RequireQnAMakerKey**: Existence of QnA Maker key

By default, these validations are all set like this:

**ForbidSpacesInProjectName**: True
**RequireBotFile**: True
**RequireEndpoints**: Production
**ForbidEndpoints**: Development
**RequireLuisKey**: True
**RequireQnAMakerKey**: True

In order to change the behavior of this validations, you can create your own properties in your `.csproj` file by adding the following code:

```.csproj
<PropertyGroup>
    <ForbidSpacesInProjectName>True</ForbidSpacesInProjectName>
    <RequireBotFile>True</RequireBotFile>
    <RequireEndpoints>Production</RequireEndpoints>
    <ForbidEndpoints>Development</ForbidEndpoints>
    <RequireLuisKey>True</RequireLuisKey>
    <RequireQnAMakerKey>True</RequireQnAMakerKey>
</PropertyGroup>
```
> **Note:** You don't need to add all of them, only the ones you want to modify.

Also, there's an option to disable the validation entirely, and for doing so, all you have to do is add the following property to your `.csproj` file:

```.csproj
<PropertyGroup>
    <DisablePublishValidation>True</DisablePublishValidation>
</PropertyGroup>
```

### CODE REVIEW

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Bot.PublishValidation
+{
+    public class ConfigurationOptions
+    {
+        public bool ForbidSpacesInProjectName { get; set; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Default to true

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        public bool RequireBotFile { get; set; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Default to true

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        public string RequiredEndpoints { get; set; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Default to "production"

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        public string ForbiddenEndpoints { get; set; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Default to string.empty

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        public bool RequireLuisKey { get; set; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Default to False

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        public bool RequireQnAMakerKey { get; set; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Default to False

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+        public ConfigurationOptions(string ForbidSpacesInProjectName, string RequireBotFile,
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Add an empty() constructor.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: //Copyright hreaders

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            this.ForbidSpacesInProjectName = ParseConfigOption(ForbidSpacesInProjectName, true);
+            this.RequireBotFile = ParseConfigOption(RequireBotFile, true);
+            this.RequireLuisKey = ParseConfigOption(RequireLuisKey, true);
+            this.RequireQnAMakerKey = ParseConfigOption(RequireQnAMakerKey, true);
+
+            this.ForbiddenEndpoints = ForbidedEndpoints;
+            this.RequiredEndpoints = RequiredEndpoints;
+        }
+
+        private bool ParseConfigOption(string configOption, bool defaultOption)
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: Please Make this a static method that returns a new ConfigurationOptions class instance. 

Leave the Configuration as a "Pure" POCO with no methods beyond the getters/setters.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs (hunk 7 above)
Comment (cleemullins) [reply libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: I would remove this ctor with all the arguments, and just have a static method do this.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/NotificationMessage.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/NotificationMessage.cs]: // Copyright

File: libraries/Microsoft.Bot.PublishValidation/Helpers/NotificationMessage.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    public enum NotificationMessageTypes
+    {
+        Warning = 0,
+        Error = 1
+    }
+
+    public class NotificationMessage
+    {
+        public string message { get; set; }
+        public int type { get; set; }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/NotificationMessage.cs]: There's an Enum above. Why have this be an int?

File: libraries/Microsoft.Bot.PublishValidation/Progam.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+    class Program
+    {
+        // Return codes
+        private const int ERROR = 2;
+        private const int OK = 0;
+
+        public static int Main(string[] args)
+        {
+            var projectPath = args[0].ToString();
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Progam.cs]: This seems... really fragile. Is there a better way to do this?

File: libraries/Microsoft.Bot.PublishValidation/README.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+>
+> RequireLuisKey: True
+>
+> RequireQnAMakerKey: True
+
+In order to change the behavior of this validations, you can create your own properties in your `.csproj` file by adding the following code:
+
+```.csproj
+<PropertyGroup>
+    <ForbidSpacesInProjectName>True</ForbidSpacesInProjectName>
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/README.md]: I don't understand how these XML methods are mapped to the configuration class. I'm clearly missing something obvious.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/NotificationMessage.cs (hunk 11 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/Helpers/NotificationMessage.cs]: We already updated this class to use the Enum.

File: libraries/Microsoft.Bot.PublishValidation/Progam.cs (hunk 12 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/Progam.cs]: We used a PowerShell-like style to receive the parameters; this means, the boolean parameters are taken as true if present and the CSV parameters are taken with the values sent; and if not present, their default value is used

File: libraries/Microsoft.Bot.PublishValidation/README.md (hunk 13 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/README.md]: In the [Microsoft.Bot.PublishValidation.targets] (/libraries/Microsoft.Bot.PublishValidation/build/Microsoft.Bot.PublishValidation.targets) there are some properties that take the values of these properties located in the `.csproj` file. We added some comments in the `.targets` file for better understanding how it works.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs (hunk 1 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: The default values were in [Microsoft.Bot.PublishValidation.targets](/libraries/Microsoft.Bot.PublishValidation/build/Microsoft.Bot.PublishValidation.targets) but had to be updated and so we moved them to the ConfigurationOption class as you suggested.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs (hunk 9 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationOptions.cs]: We created the class `ConfigurationParser` with the static method `ParseConfiguration`, that returns a new ConfigurationOptions class instance.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationParser.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+// Licensed under the MIT License.
+
+namespace Microsoft.Bot.PublishValidation
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    public class ConfigurationParser
+    {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationParser.cs]: Public static class, to enforce that there are only static methods.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationParser.cs [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            try
+            {
+                var configurationOptions = new ConfigurationOptions
+                {
+                    // Parse the PROJECT PATH
+                    ProjectPath = options.Contains(PROJECTPATH) ? options[options.ToList().IndexOf(PROJECTPATH) + 1] : string.Empty,
+
+                    // If the option 'AllowSpacesInProjectName' is present, the process wont validated if the project's name has white spaces
+                    ForbidSpacesInProjectName = options.Contains(ALLOWSPACESINPROJECTNAME) ? false : true,
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationParser.cs]: grammer on comment. Spelling on (wont -> won't)

File: libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationParser.cs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+                    // Check if the configuration will require the QNA MAKER KEY
+                    RequireQnAMakerKey = options.Contains(REQUIREQNAMAKERKEY) ? true : false
+                };
+
+                return configurationOptions;
+            }
+            catch (Exception ex)
+            {
+                throw ex;
+            }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/ConfigurationParser.cs]: Not clear on why this try/catch is here. 

As it sits, it's incorrect. If you're using this just to set a breakpoint, then it should be "throw" not "throw ex".

File: libraries/Microsoft.Bot.PublishValidation/README.md (hunk 13 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/README.md]: In the [Microsoft.Bot.PublishValidation.targets] (/libraries/Microsoft.Bot.PublishValidation/build/Microsoft.Bot.PublishValidation.targets) there are some properties that take the values of these properties located in the `.csproj` file. We added some comments in the `.targets` file for better understanding how it works.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+                // Load the first .bot file from the provided folder.
+                // Also validates its existence (throws an exception is there isn't any file)
+                var botConfiguration = BotValidatorHelper.LoadFromFolder(configurationOptions.ProjectPath);
+
+                return BotValidatorHelper.ValidateBotFile(botConfiguration, configurationOptions, ref messages);
+            }
+            catch (Exception ex)
+            {
+                throw ex;
+            }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs]: Similar comment. Not sure why there is a try/catch here. If we want to keep it (handy for breakpoints) then change the "throw ex" to just "throw" so the exception isn't munched.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+
+        /// <summary>
+        /// Performs the validation of the specified .bot file, according the specified validation options
+        /// </summary>
+        /// <param name="botConfiguration"></param>
+        /// <param name="options"></param>
+        /// <param name="messages"></param>
+        /// <returns></returns>
+        private static bool ValidateBotFile(BotConfiguration botConfiguration, ConfigurationOptions options, ref IEnumerable<NotificationMessage> messages)
+        {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs]: Not clear why this is a ref parameter, or an IEnumerable. 

Seems like this should just be a List<NotificationMessage>, and passes as a standard parameter.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+                        validationResult = false;
+                    }
+                }
+
+                return validationResult;
+            }
+            catch (Exception ex)
+            {
+                throw ex;
+            }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs]: Same comment(s)

File: libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs [hunk 20]
```diff
@@ -0,0 +1,377 @@
...
+            {
+                missingEndpoints = string.Empty;
+                return false;
+            }
+
+            // Checks that all the specified endpoints are/aren't in the bot file
+            foreach (var endpoint in endpoints)
+            {
+                if (botEndpoints.Any(ep => ((EndpointService)ep).Name.Trim() == endpoint.Trim()) != required)
+                {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs]: I can't follow this if statements - too complex. Can you simplify?

File: libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs [hunk 21]
```diff
@@ -0,0 +1,377 @@
...
+                    missingEndpointsList.Add(endpoint.Trim());
+                }
+            }
+
+            // If there is at least one missing endpoint, the method will return an error message listing all of them
+            if (missingEndpointsList.Count() != 0)
+            {
+                missingEndpoints = string.Join(", ", missingEndpointsList);
+                //missingEndpoints = "\n\t*" + missingEndpoints;
+                return false;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs]: REmove dead code

File: libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs (hunk 20 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs]: We've divided that line into a few more and also added comments.

File: libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs (hunk 18 above)
Comment (GasparAcevedoZainSouthworks) [reply libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs]: We already updated this method to use a List.

**Comment by GasparAcevedoZainSouthworks** [comment]
We've applied the requested changes. Also, We've created a new pull-request (PR [#1225](https://github.com/Microsoft/botbuilder-dotnet/pull/1225)) where we apply some linting rules to this project's files (`Microsoft.Bot.PublishValidation`) and, because of it, must be merged after this PR.

**Comment by cleemullins** [comment]
@GasparAcevedoZainSouthworks Once we've cut the 4.2 release branch and bits, I'll look at merging this in. Timing wise, it looks like it'll be part of the 4.3 cycle.

**Comment by cleemullins** [comment]
@GasparAcevedoZainSouthworks Can you move this code to a branch off the repo? I would really like our automated build process to pick it up. Probably easiest to just update the PR to merge into a different branch, then submit a PR to merge that into master.

**Comment by GasparAcevedoZainSouthworks** [comment]
> @GasparAcevedoZainSouthworks Can you move this code to a branch off the repo? I would really like our automated build process to pick it up. Probably easiest to just update the PR to merge into a different branch, then submit a PR to merge that into master.

@cleemullins, We have updated this PR to merge with the branch **_southworks/before-publish-validations_** and once it's merged, we will create another PR to merge that with **_master_** branch. We also have performed a similar action with the #1228, changing the target branch from master to **_southworks/before-publish-validations-specific-botfile_**, and, once it's merged, a new PR targeting master will be created. We think that it might be more useful to use the #1228 instead of the #1214, because it's more complete, and also for avoiding one merge.

**Comment by cleemullins** [comment]
Sadly, closing this for now. With the passing of .bot files, we may revisit this.

**Pull request closed without merging**